backend/src/worker.rs

```diff
@@ -1351,175 +1361,228 @@ async fn handle_child(
+                          SET canceled = true
+                            , canceled_by = 'timeout',
+                            , canceled_reason = $1
+                        WHERE id = $2
+                    r"#,
+                )
+                .bind(format!("duration > {}", timeout.as_secs()))
+                .bind(job_id)
+                .execute(&db)
+                .await
+                {
+                    tracing::error!(%job_id, %err, "error setting cancelation reason");
                 }
-            } => {
-                child.kill().await?;
-                return Err(Error::ExecutionErr("execution interrupted".to_string()).into())
             }
         };
-        r
-    });
-
-    let (tx, mut rx) = mpsc::channel::<String>(100);
-
-    tokio::spawn(async move {
-        while !done4.load(Ordering::Relaxed) {
-            let send = tokio::select! {
-                Ok(Some(out)) = reader.next_line() => {
-                    if out.len() > MAX_LOG_SIZE as usize {
-                        tracing::info!("Line is too big");
-                        let _ = tx.send(format!("Line is too big")).await;
-                        done4.store(true, Ordering::Relaxed);
-                        break;
-                    } else {
-                        tx.send(out).await
+
+        /* send SIGKILL and reap child process */
+        let (_, kill) = futures::future::join(set_reason, child.kill()).await;
+        kill.map(|()| None)
+    };
+
+    /* a future that reads output from the child and appends to the database */
+    let lines = async move {
+        /* log_remaining is zero when output limit was reached */
+        let mut log_remaining = (MAX_LOG_SIZE as usize).saturating_sub(logs.chars().count());
+        let mut result = io::Result::Ok(());
+        let mut output = output;
+        let mut write: OptionFuture<_> = None.into();
+
+        while let Some(line) = output.by_ref().next().await {
+            /* after receiving a line, continue until some delay has passed
+             * _and_ the previous database write is complete */
+            let until_write = write.clone();
+            let until = sleep(write_logs_delay).then(|()| until_write);
+            let mut read_lines = ready(line)
+                .into_stream()
+                .chain(output.by_ref())
+                .take_until(until)
+                .boxed();
+
+            /* Read up until an error is encountered,
+             * handle log lines first and then the error...
+             *
+             * (This looks a bit nicer using try_for_each but the side-effects/capturing
+             * in FnMut closure seems impractical, _maybe_ a futures::Sink would work but
+             * I know next to nothing about that.) */
+            let mut joined = String::new();
+
+            while let Some(line) = read_lines.next().await {
+                match line {
+                    Ok(_) if log_remaining == 0 => (),
```

Comment:
Is that necessary if you break on log_remaining  == 0 in the case below ?

Reply:
It doesn't break when the read limit is reached, only when there is a read error.

But maybe it should break. A reason to continue reading is so the job process doesn't get a broken pipe while writing to stdout/stderr, but since we're killing it anyway maybe it doesn't matter. Although then it would fail with a non-zero status code and we would report that as a different failure.

Reply:
I think it should break since the mechanism is intended to be a protection against "abuse"

Reply:
It already sets a flag to kill the process.